packages/paima-sdk/paima-events/src/app-events.ts

```diff
@@ -0,0 +1,200 @@
+import type { LogEvent, LogEventFields } from './types.js';
+import type { genEvent } from './types.js';
+import { toPath, TopicPrefix } from './types.js';
+import type { TSchema, Static } from '@sinclair/typebox';
+import { Value } from '@sinclair/typebox/value';
+import { keccak_256 } from 'js-sha3';
+
+export function toSignature<T extends LogEvent<LogEventFields<TSchema>[]>>(event: T): string {
+  return event.name + '(' + event.fields.map(f => f.type.type).join(',') + ')';
+}
+
+export function toTopicHash<T extends LogEvent<LogEventFields<TSchema>[]>>(event: T): string {
+  return keccak_256(toSignature(event));
+}
+
+// this generates the type for the events import.
+export const generateAppEvents = <
+  const T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>,
+>(
+  entries: T
+): {
+  [K in T[number] as K['name']]: (ReturnType<typeof toPath<K, typeof TopicPrefix.App>> & {
+    definition: T[0];
+    topicHash: string;
+  })[];
+} => {
+  let result: Record<string, any> = {}; // we can't know the type here
+  for (const event of entries) {
+    if (!result[event.name]) {
+      result[event.name] = [];
+    }
+
+    result[event.name].push({
+      ...toPath(TopicPrefix.App, event),
+      // keep the original definition around since it's nicer to work with, it
+      // also has the advantage that it allows recovering the initial order in
+      // case the signature/topicHash needs to be computed again, which can't be
+      // done from the path (since you don't know which non indexed fields go in
+      // between each indexed field).
+      definition: event,
+      // we add this to avoid having to re-compute it all the time.
+      topicHash: toTopicHash(event),
+    });
+  }
+  return result as any;
+};
+
+// create payload for the stf from an object.
+export function createEventForStf<T extends ReturnType<typeof genEvent>>(
```

Comment:
This can be used like this in the stf function, for example:

```ts
const event = createEventForStf(user, StfEvent1, {
    tag: 'command1',
    data: 'test',
});

events.push(event);

return { stateTransitions: [], events  };
```

I'm open to suggestions for a better name for it though.

The main annoying part about this is function is that you need to name the overloaded events with some sort of unique suffix, and use the correct index.

Reply:
> The main annoying part about this is function is that you need to name the overloaded events with some sort of unique suffix, and use the correct index.

It's not clear to me why this would be required. This isn't required in Ethereum, so why wold it be required for us?

Reply:
I don't mean the name of the event itself, I mean the name of the javascript variable that has the definition. Or if you define them directly in an array, then you need the index in the array.

Although this is mostly because the topic is part of the input, and you need to know which overload it is (see this comment https://github.com/PaimaStudios/paima-engine/pull/409#discussion_r1719369411) to be able to provide it.

If we removed that requirement then I think we could have a simpler api. Although in that case we need to compute it from the data inside the engine, so that we know which type it is.